### PR TITLE
JexlNodes negate builds proper reference expression

### PR DESCRIPTION
**Pull request by apmoriarty**
General clean up of how negations are handled
- JexlNodes negate call now builds proper reference-referenceExpression chain, to include marker nodes
- PushdownNegations dropNot call now drops all relevant nodes to the negation

### CODE REVIEW

**Review by ivakegg — DISMISSED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PullupUnexecutableNodesVisitor.java [hunk 1]
```diff
@@ -245,6 +245,10 @@ public Object visit(ASTReference node, Object data) {
         QueryPropertyMarker.Instance instance = QueryPropertyMarker.findInstance(node);
         if (instance.isType(ASTDelayedPredicate.class)) {
             JexlNode source = ASTDelayedPredicate.unwrapFully(node, ASTDelayedPredicate.class);
+            // when pulling up a delayed marker that is negated, a reference expression must be persisted
+            if (!(source instanceof ASTReferenceExpression) && JexlNodes.findNegatedParent(node)) {
```
Comment (lbschanno) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PullupUnexecutableNodesVisitor.java]: It is possible for the source to be an instance of ASTReference, do you need to account for this here?

**Review by lbschanno — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PullupUnexecutableNodesVisitor.java (hunk 1 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PullupUnexecutableNodesVisitor.java]: Addressed this by always dereferencing the source and then creating an `ASTReferenceExpression` when the source is an `ASTOrNode`, an `ASTAndNode`, or if the source is negated.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PullupUnexecutableNodesVisitor.java [hunk 2]
```diff
@@ -245,7 +250,12 @@ public Object visit(ASTReference node, Object data) {
         QueryPropertyMarker.Instance instance = QueryPropertyMarker.findInstance(node);
         if (instance.isType(ASTDelayedPredicate.class)) {
             JexlNode source = ASTDelayedPredicate.unwrapFully(node, ASTDelayedPredicate.class);
-            JexlNodes.swap(node.jjtGetParent(), node, source);
+            source = JexlASTHelper.dereference(source);
+            // when pulling up a delayed marker that is negated, a reference expression must be persisted
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PullupUnexecutableNodesVisitor.java:253]: I am wondering if this dereference and wrap if an and/or should be a JexlASTHelper method: JexlASTHelper.wrapIfNeeded(node)

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownNegationVisitor.java [hunk 3]
```diff
@@ -231,19 +234,47 @@ public static JexlNode applyDeMorgans(JexlNode root, boolean negateRoot) {
...
+     */
     private JexlNode dropNot(ASTNotNode toDrop, NegationState state) {
-        JexlNode newNode = toDrop.jjtGetChild(0);
+        JexlNode child = toDrop.jjtGetChild(0);
         
         // no need to swap nodes if toDrop is the root
         if (toDrop.jjtGetParent() != null) {
+            
+            // check for adjacent not nodes
+            if (child instanceof ASTNotNode || QueryPropertyMarker.findInstance(child).isAnyType()) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownNegationVisitor.java:253]: What if the child of the child is a Not Node?  What if the child of the child of the child is a NotNode?  Doesn't the visitor pattern do the appropriate thing as it moves down the tree?

File: warehouse/query-core/src/main/java/org/apache/commons/jexl2/parser/JexlNodes.java [hunk 4]
```diff
@@ -170,8 +171,19 @@ public static <T> void setLiteral(ASTStringLiteral literal, String value) {
...
+    /**
+     * Negate the provided JexlNode
+     *
+     * @param node
+     *            an arbitrary JexlNode
+     * @return a negated version of the provided JexlNode
+     */
     public static ASTNotNode negate(JexlNode node) {
-        return children(new ASTNotNode(ParserTreeConstants.JJTNOTNODE), wrap(node));
+        if (QueryPropertyMarker.findInstance(node).isAnyType()) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/org/apache/commons/jexl2/parser/JexlNodes.java:182]: We could use that wrapIfNeeded call I suggested above.

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownNegationVisitor.java (hunk 3 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PushdownNegationVisitor.java:253]: This is handled by the calling method. After dropping the negation it immediately visits the returned node. I believe several of the tests for this visitor cover the exact case you're thinking of.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PullupUnexecutableNodesVisitor.java (hunk 2 above)
Comment (apmoriarty) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/PullupUnexecutableNodesVisitor.java:253]: I can extract that logic to a local method. I'm not thinking of any broader use cases for such a method outside of this visitor. If you can think of any I'm happy to move that logic to the JexlASTHelper.

**Review by FineAndDandy — APPROVED** [review]

**Review by ivakegg — APPROVED** [review]

**Review by lbschanno — APPROVED** [review]

**Pull request merged**